csrc/optimization/optimization_pass.h

```diff
@@ -0,0 +1,85 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <ir/interface_nodes.h>
+
+#include <mutex>
+
+namespace nvfuser::optimization {
+
+using FusionPass = std::function<void(Fusion*)>;
+
+//! [experimental API]
+//! Base class to unify optimization pass APIs.
+//! OptimizationPass is functional and defines the granularity of mutation
+//! passes that is used to compose OptimizationGroups
+class TORCH_CUDA_CU_API OptimizationPass {
+ public:
+  virtual void run(Fusion*) = 0;
+  virtual std::string name() = 0;
+  virtual ~OptimizationPass() = default;
+};
+
+//! [experimental API]
+//! Base class to unify optimization group APIs.
+//! OptimizationGroup composes optimization passes that is used at certain stage
+//! in the runtime system. OptimizationGroup can be turned on/off
+//! programmatically with the `setEnabled/flipEnabled` API. There's helper
+//! template OptimizationGroupGuard to temporarily switch the enablement within
+//! the context. Note the we are using a curiously recurring template pattern
+//! here to ensure that static objects are unique for each DerivedClass. In
+//! order to apply OptimizationGroup with the switch enabled, you need to run
+//! the function with `OptimizationGroup<DerivedClass>::runPass(...)`
+template <typename DerivedClass>
+class TORCH_CUDA_CU_API OptimizationGroup {
+ public:
+  static bool flipEnabled(bool flip) {
+    static std::mutex mutex_;
+    static bool enable_flag_ = true;
```

Comment:
Is this some know design pattern? Would it make sense to make `mutex_` and `enable_flat_` a static member of the class, so that we can directly read it? It looks strange to me that we use `flipEnabled` to read the value of `enabled_flat_`

Reply:
And what is the point of protecting with mutex? If all we need is the functionality of read it, set `true` or set `false`, do we need atomic behavior? And should we instead use thread local variables to work better with async compilation?

Reply:
Yes -> `curiously recursive template`.
1. making `mutex_` and `enable_flat_` as a static member of the class requires inherited class to define them. Which is more code. I'm trying to keep the entire logic inside `OptimizeationGroup`, so inherited classes doesn't need to do anything extra to have the switch working.
1.1. Re: `flipEnabled`. lol it just feels straightforward to me with that, I can change this to a setter/getter methods if this is confusing to use.

2. Re: protecting with mutex, if we are reading and writing it in mutithread condition, we do need to protect it right? Arguably with our python API, you can say thread safety isn't a real concern. But then, it doesn't hurt neither :)
2.1 Re: thread_local vs static, this is mostly a design choice. I don't think one is better than the other. It boils down to how `framework` works. i.e. Doesn't PyTorch backward works on a separate thread, so if this is indeed a thread_local flag, it gets tricky on how to turn off these passes on backwards?
I think having a static/global flag makes debugging easier in general.

Reply:
I think as long as you define your static members as `inline`, you won't have the issue of having to ask derived class to define them? This is a new feature of C++17, see 
https://cpp.sh/?source=%2F%2F+Example+program%0A%23include+%3Ciostream%3E%0A%23include+%3Cstring%3E%0A%0Aint+main()%0A%7B%0A++std%3A%3Astring+name%3B%0A++std%3A%3Acout+%3C%3C+%22What+is+your+name%3F+%22%3B%0A++getline+(std%3A%3Acin%2C+name)%3B%0A++std%3A%3Acout+%3C%3C+%22Hello%2C+%22+%3C%3C+name+%3C%3C+%22!%5Cn%22%3B%0A%7D

Reply:
And if we don't have to define them inside one function, I think a getter/setter makes more sense than flip.

Reply:
And I also think with the current implementation, the `mutex` protects nothing. For example, in `setEnabled`, this is what can happen:
```C++
  // user call this method with enabled = true
  static bool setEnabled(bool enabled) {
    // enable_flag_ is false
    auto tmp = flipEnabled(false);
    // tmp is false, however, after we read tmp, another thread change enable_flag_ to true
    if (enabled != tmp) { // the condition evaluates to true, so will flip
      flipEnabled(true);
    }
    // after flip, we now have enable_flag_ = false, this is not what we want
    return tmp;
  }
```

Reply:
If the mutex does make this flag thread-safe, then I don't want to lose this safety just to save a mutex, but this mutex is actually protecting nothing, so why bother having it.

Reply:
I think we should be clear about the thread-safety model here? Do we want to be thread-safe for this flag? If the answer is yes, then what is the correct behavior if one thread set this flag as true and the other flag set it as false? Do we want these two threads to compete on the value of this flag? For example, if gtest provide the functionality of running tests in multiple thread, do we want the disable guard in one test change the behavior of the another test? And if the answer is no, then we shouldn't have this mutex at all.

Reply:
Good catch on the threading bug I have!

I think what you are saying makes sense to me. i.e. a global thing with a mutex guard feels clumsy and pretty useless... I'll back out from this and just make the flag an `atomic` instead.